crypto/evp/evp_enc.c

```diff
@@ -336,7 +354,20 @@ int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
             return 0;
         }
     }
+
+    /*
+     * Aliasing in and out in this way is counter intuitive and only supported
+     * for legacy applications.
+     * Don't rely on this to stay the same in the next major version.
```

Comment:
This is somewhat confusing comment... What is it that is suggested to omit? I'd argue that following should always work:
```
    unsigned char *inp = buf, *out = buf;

    for (i = 0; i < sizeof(buf); i++) {
        EVP_EncryptUpdate(ctx, out, &outl, inp++, 1);
	out += outl;
    }
```
In other words "pure" in-place. Note that currently it's effectively handled by `(PTRDIFF_T)in != (PTRDIFF_T)(out + i)` rather than by is_any_overlapping...

Reply:
I don't understand why this is not written as

`EVP_EncryptUpdate(ctx, out, &outl, inp, sizeof(buf));
`

Reply:
Because question is not what it right thing to do in the snippet, but what should work no matter what. If you need something that helps imagining, consider *inp = readioport(); prior call to EVP_EncryptUpdate.

Reply:
But in that use case, why should you write char-after-char in the same buffer space where the cipher text is
constructed?
Why not do simply `char ch = readioport(); EVP_EncryptUpdate(ctx, out, &outl, &ch, 1);` ?

Reply:
The real question here is, if I read the documentation of EVP_EncryptUpdate:
```
EVP_EncryptUpdate() encrypts B<inl> bytes from the buffer B<in> and
writes the encrypted version to B<out>. This function can be called
multiple times to encrypt successive blocks of data. The amount
of data written depends on the block alignment of the encrypted data:
as a result the amount of data written may be anything from zero bytes
to (inl + cipher_block_size - 1) so B<out> should contain sufficient
room. The actual number of bytes written is placed in B<outl>. It also
checks if B<in> and B<out> are partially overlapping, and if they are
0 is returned to indicate failure.
```

I don't see how _partially overlapping_ can be understood to allow the
aliasing that happens in this loop:

```
unsigned char *inp = buf, *out = buf;
for (i = 0; i < sizeof(buf); i++) {
        EVP_EncryptUpdate(ctx, out, &outl, inp++, 1);
	out += outl;
    }
```
How can one consider `EVP_EncryptUpdate(ctx, out, &outl, out+1, 1);` a partial overlapping buffer?
but not `EVP_EncryptUpdate(ctx, out, &outl, out, 1);` in the second iteration of the loop, (when ctx->buf_len == 1) ?

Reply:
> Why not do simply

I'm perfectly aware of what one can do [better], but thing is that you can't assume that you'll be in position to pose this question to each user. Instead you have to assume that caller is not as rational as you :-) In other words it doesn't matter how irrational suggested snippet is, it should just work...

Reply:
Keep in mind difference between partially and fully overlapping. Fully overlapping, or in other words in-place processing, should work. [I have to step out now, i.e. more detailed remark will be provided later.]

Reply:
> How can one consider EVP_EncryptUpdate(ctx, out, &outl, out+1, 1); a partial overlapping buffer? but not EVP_EncryptUpdate(ctx, out, &outl, out, 1); in the second iteration of the loop, (when ctx->buf_len == 1) ?

Well, there is no good formal answer to this question. But I'd say that this is the case when *intention* counts. And above mentioned second iteration of the loop is arguably indication of *legitimate* intention. I.e. if you see that end of [tp be] written data aligns perfectly with input, you know you are looking at in-place processing.

[BTW, one can probably argue that first case should work too. Well, as long as caller always passes one byte at a time (or small enough) and advances input pointer, so that complete blocks are written one by one and never overstep the input. But I'd say that we are free to make the choice to claim this case unsupported.]

Reply:
All depends on the definition of partially overlapping.
Our doc fails to define it because it should be self-evident.

My definition would be:
[a..b] and [c..d] are partially overlapping IF the intersection
of both intervals is not empty, and there exists at least one
element in [a..b] that is not in [c..d] or vice-versa.

But the current behaviour of EVP_EncryptUpdate with regard to partially
overlapping in and out buffers is in fact _totally different_ from
what the doc says.  No matter what I write in this comment.
I do just not like to encourage people to use this kind of programming style.
Which is very hard to explain and it is fragile, because if you
load just one byte in the internal buffer before the loop starts,
the loop will fail immediately.